pkg/distributor/forwarding/forwarding.go

```diff
@@ -311,7 +349,36 @@ func (f *forwarder) splitByTargets(targetEndpoint string, tsSliceIn []mimirpb.Pr
 		}
 	}
 
-	return tsToIngest, tsByTargets
+	return tsToIngest, tsByTargets, err
+}
+
+// samplesNeedFiltering takes a slice of samples and a timestamp before which samples should be filtered out,
+// it returns true if some samples are older than the given timestamp or false otherwise.
+// It assumes that the samples are sorted by timestamp.
+func samplesNeedFiltering(samples []mimirpb.Sample, dontForwardBefore int64) bool {
+	if len(samples) == 0 {
+		return false
+	}
+
+	return samples[0].TimestampMs < dontForwardBefore
+}
+
+// filterSamplesBefore filters a given slice of samples to only contain samples that have timestamps newer or equal to
+// the given timestamp. It relies on the samples being sorted by timestamp.
+// If some samples have been filtered the second return value is true, otherwise it is false.
+func filterSamplesBefore(samples []mimirpb.Sample, dontForwardBefore int64) []mimirpb.Sample {
+	if dontForwardBefore == 0 {
+		return samples
+	}
+
+	for sampleIdx, sample := range samples {
+		if sample.TimestampMs >= dontForwardBefore {
+			// In most cases the first sample should already meet this condition and we can return quickly.
+			return samples[sampleIdx:]
+		}
+	}
+
+	return samples[:0]
```

Comment:
why don't we return `nil` in this case?

Reply:
I don't think it makes a difference really, we can also return `nil`

Reply:
changed here 4ab289574